### PR TITLE
Fix getBBMemoryAccess

### PR DESCRIPTION
**Pull request by nsurbay**
Used the instID of ShadowInfo to search the shadow in a Sequence. The fix avoid to return no shadow when the sequence is a split of another.

### CODE REVIEW

File: src/Engine/Engine.h [hunk 1]
```diff
@@ -201,11 +201,12 @@ class Engine {
     /*! Add a custom instrumentation rule to the engine. Requires internal headers
      *
      * @param[in] rule A custom instrumentation rule.
+     * @param[in] top_list Place the rule before all existed rules of it category
```
Comment (romainthomas) [inline src/Engine/Engine.h:204]: Place the rule before all **existing** rules of it category of **its** category

File: src/Engine/Engine.cpp [hunk 2]
```diff
@@ -410,16 +410,24 @@ bool Engine::run(rword start, rword stop) {
...
-uint32_t Engine::addInstrRule(InstrRule rule) {
+uint32_t Engine::addInstrRule(InstrRule rule, bool top_list) {
     uint32_t id = instrRulesCounter++;
     RequireAction("Engine::addInstrRule", id < EVENTID_VM_MASK, return VMError::INVALID_EVENTID);
     blockManager->clearCache(rule.affectedRange());
     switch(rule.getPosition()) {
         case InstPosition::PREINST:
-            instrRules.insert(instrRules.begin(), std::make_pair(id, (InstrRule::SharedPtr) rule));
+            if (top_list) {
+                instrRules.push_back(std::make_pair(id, (InstrRule::SharedPtr) rule));
```
Comment (romainthomas) [inline src/Engine/Engine.cpp:420]: Prefer ``emplace_back()``

File: src/Engine/Engine.cpp [hunk 3]
```diff
@@ -410,16 +410,24 @@ bool Engine::run(rword start, rword stop) {
...
+            } else {
+                instrRules.insert(instrRules.begin(), std::make_pair(id, (InstrRule::SharedPtr) rule));
+            }
             break;
         case InstPosition::POSTINST:
-            instrRules.push_back(std::make_pair(id, (InstrRule::SharedPtr) rule));
+            if (top_list) {
+                instrRules.insert(instrRules.begin(), std::make_pair(id, (InstrRule::SharedPtr) rule));
+            } else {
+                instrRules.push_back(std::make_pair(id, (InstrRule::SharedPtr) rule));
```
Comment (romainthomas) [inline src/Engine/Engine.cpp:429]: Prefer ``emplace_back()``

File: src/Engine/Engine.cpp [hunk 4]
```diff
@@ -410,16 +410,24 @@ bool Engine::run(rword start, rword stop) {
...
 
-uint32_t Engine::addInstrRule(InstrRule rule) {
+uint32_t Engine::addInstrRule(InstrRule rule, bool top_list) {
     uint32_t id = instrRulesCounter++;
     RequireAction("Engine::addInstrRule", id < EVENTID_VM_MASK, return VMError::INVALID_EVENTID);
     blockManager->clearCache(rule.affectedRange());
     switch(rule.getPosition()) {
         case InstPosition::PREINST:
-            instrRules.insert(instrRules.begin(), std::make_pair(id, (InstrRule::SharedPtr) rule));
+            if (top_list) {
```
[the remark this answers is not in the document]
Comment (nsurbay) [reply src/Engine/Engine.cpp:419]: We apply the InstrRule in the order of `instrRules` (see `Engine::instrument` in Engine.cpp). During the `InstrRule::instrument` (https://github.com/QBDI/QBDI/blob/master/src/Patch/InstrRule.cpp#L109), a PREINST is prepend to the current patch.

If we want to place the PREINST rule before any other PREINST rules in the patch, it must be at the end of `instrRules`.

File: src/Engine/Engine.cpp (hunk 2 above)
Comment (nsurbay) [reply src/Engine/Engine.cpp]: ack + add static_cast

**Pull request merged**